File: clients/godot-app/scripts/NetDelay.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Colyseus;

namespace Playground
{
    /// <summary>
    /// Latency injector (APPS_PLAN §3). The web playground gets delay/jitter from
    /// the JS SDK's debug panel; on localhost every other SDK needs its own, or
    /// labs 00/01/03 demonstrate nothing.
    ///
    /// This wraps a room's <see cref="Connection"/> in place, the way the Lua
    /// client wraps `connection.emit`: capture <see cref="Connection.Dispatch"/>
    /// and <see cref="Connection.Transmit"/>, put a queue in front, and call the
    /// captured pair when a packet comes due. One seam each way, no subclass, and
    /// nothing global — each room gets its own injector.
    ///
    /// Two properties are load-bearing:
    ///
    ///  - No reordering: each packet's deliver-at is clamped to ≥ the previous
    ///    one's, because the wire is a stream and TCP never reorders (the JS
    ///    debug panel does the same).
    ///  - The wrap happens AFTER the join, so the handshake is never delayed —
    ///    only gameplay traffic is. That is what a real link looks like from the
    ///    room's point of view, and it means awaiting a join can't deadlock on a
    ///    queue nobody is draining yet.
    /// </summary>
    public class NetDelay
    {
        /// <summary>
        /// A queued packet. It carries the payload rather than a closure over it
        /// on purpose: a `() => inner(e)` per frame would be two heap objects per
        /// message, and this queue sees every patch and every input.
        /// </summary>
        private struct Packet
        {
            public double DeliverAt;
            public ConnectionEvent Event;   // inbound
            public byte[] Data;             // outbound
        }

        public static double DelayMs;
        public static double JitterMs;

        private static readonly List<NetDelay> Live = new List<NetDelay>();
        private static readonly Random Rng = new Random(0x5EED);

        private readonly Queue<Packet> _inbound = new Queue<Packet>();
        private readonly Queue<Packet> _outbound = new Queue<Packet>();
        private double _inLast, _outLast;
        private readonly object _lock = new object();
        private readonly Connection _conn;
        private Action<ConnectionEvent> _innerDispatch;
        private Func<byte[], Task> _innerTransmit;

        private NetDelay(Connection conn) { _conn = conn; }

        /// <summary>
        /// Put an injector in front of a joined room's socket. Idempotent: a room
        /// that is already wrapped keeps its existing queue.
        /// </summary>
        public static NetDelay Wrap(Connection conn)
        {
            if (conn == null) return null;
            lock (Live)
            {
                foreach (var w in Live) if (w._conn == conn) return w;
            }

            var nd = new NetDelay(conn);
            nd._innerDispatch = conn.Dispatch;
            nd._innerTransmit = conn.Transmit;

            // Both lambdas are allocated once, here — not per packet.
            conn.Dispatch = nd.QueueInbound;
            conn.Transmit = nd.QueueOutbound;

            Live.Add(nd);
            return nd;
        }

        public static void SetLatency(double delayMs, double jitterMs)
        {
            DelayMs = Math.Max(0, delayMs);
            JitterMs = Math.Max(0, jitterMs);
        }

        /// <summary>A named delay/jitter pair — what the `L` key cycles through.</summary>
        public readonly struct Preset
        {
            public readonly double Delay, Jitter;
            public readonly string Label;
            public Preset(double delay, double jitter, string label)
            {
                Delay = delay; Jitter = jitter; Label = label;
            }
        }

        public static readonly Preset[] Presets =
        {
            new Preset(0, 0, "off"),
            new Preset(80, 10, "80 ms + 10 jitter"),
            new Preset(200, 0, "200 ms"),
            new Preset(200, 80, "200 ms + 80 jitter"),
            new Preset(400, 60, "400 ms + 60 jitter"),
        };

        public static int PresetIndex { get; private set; }
        public static string PresetLabel => Presets[PresetIndex].Label;

        public static void UsePreset(int index)
        {
            PresetIndex = ((index % Presets.Length) + Presets.Length) % Presets.Length;
            SetLatency(Presets[PresetIndex].Delay, Presets[PresetIndex].Jitter);
        }

        public static void NextPreset() => UsePreset(PresetIndex + 1);

        /// <summary>Forget every socket and zero the latency — between test cases.</summary>
        public static void Reset()
        {
            Live.Clear();
            DelayMs = JitterMs = 0;
            PresetIndex = 0;
        }

        /// <summary>Kill every live socket, so the SDK sees a drop and not a leave.</summary>
        public static void DropAll()
        {
            for (int i = 0; i < Live.Count; i++) Live[i]._conn.Drop();
        }

        public static int InFlight()
        {
            int n = 0;
            for (int i = 0; i < Live.Count; i++)
            {
                var w = Live[i];
                lock (w._lock) n += w._inbound.Count + w._outbound.Count;
            }
            return n;
        }

        /// <summary>
        /// Drain everything due. Call once per frame from Update().
        ///
        /// Indexed, not foreach-over-a-copy: `Live` is only ever mutated on the
        /// main thread (Wrap runs on an awaited continuation, Reset from a test),
        /// while the per-socket queues are what the socket thread touches — and
        /// those have their own lock. Copying the list here would allocate an
        /// array every frame for nothing.
        /// </summary>
        public static void PumpAll()
        {
            double now = RoomClock.GetNow();
            for (int i = 0; i < Live.Count; i++) Live[i].Pump(now);
        }

        // Rng is shared across connections but Enqueue only holds the per-socket lock.
        private static double OneWay()
        {
            lock (Rng) return DelayMs + Rng.NextDouble() * JitterMs;
        }

        private void QueueInbound(ConnectionEvent e)
        {
            double at = RoomClock.GetNow() + OneWay();
            lock (_lock)
            {
                if (at < _inLast) at = _inLast;    // the wire never reorders
                _inLast = at;
                _inbound.Enqueue(new Packet { DeliverAt = at, Event = e });
            }
        }

        private Task QueueOutbound(byte[] data)
        {
            double at = RoomClock.GetNow() + OneWay();
            lock (_lock)
            {
                if (at < _outLast) at = _outLast;
                _outLast = at;
                _outbound.Enqueue(new Packet { DeliverAt = at, Data = data });
            }
            return Task.CompletedTask;
        }

        private void Pump(double now)
        {
            while (true)
            {
                Packet p;
                lock (_lock)
                {
                    if (_outbound.Count == 0 || _outbound.Peek().DeliverAt > now) break;
                    p = _outbound.Dequeue();
                }
                _innerTransmit(p.Data);       // outside the lock: this hits the socket
            }
            while (true)
            {
                Packet p;
                lock (_lock)
                {
                    if (_inbound.Count == 0 || _inbound.Peek().DeliverAt > now) break;
                    p = _inbound.Dequeue();
                }
                _innerDispatch(p.Event);      // outside the lock: this reaches the room
            }
        }
    }
}

File: clients/defold-app/playground/net_delay.lua
--
-- Latency injector (APPS_PLAN §3). The web playground gets delay/jitter from the
-- JS SDK's debug panel; on localhost every other SDK needs its own, or labs
-- 00/01/03 demonstrate nothing.
--
-- Lua has the cleanest seam of the four. `Connection` is an EventEmitter, so
-- replacing its `emit` puts us in front of EVERY listener at once — including
-- ones registered later. Subscribing instead would only run us alongside the
-- room's handler, which cannot delay anything. Outbound goes through `send`.
--
-- Two properties are load-bearing:
--
--  - No reordering: each packet's deliver-at is clamped to >= the previous
--    one's, because the wire is a stream and TCP never reorders (the JS debug
--    panel does the same).
--  - "open" is delayed with the rest. `Connection:_connect` flips `state` to
--    OPEN synchronously, so sends still reach our queue — but the room must not
--    learn it joined before the frames that arrived with it.
--
local M = {
  delay_ms = 0,
  jitter_ms = 0,
  _live = {},
  _seed = 0x5EED,
}

--- Named delay/jitter pairs — what the `L` key cycles through.
M.PRESETS = {
  { delay = 0,   jitter = 0,  label = "off" },
  { delay = 80,  jitter = 10, label = "80 ms + 10 jitter" },
  { delay = 200, jitter = 0,  label = "200 ms" },
  { delay = 200, jitter = 80, label = "200 ms + 80 jitter" },
  { delay = 400, jitter = 60, label = "400 ms + 60 jitter" },
}
M.preset_index = 1

function M.set_latency(delay_ms, jitter_ms)
  M.delay_ms = math.max(0, delay_ms or 0)
  M.jitter_ms = math.max(0, jitter_ms or 0)
end

function M.use_preset(index)
  M.preset_index = ((index - 1) % #M.PRESETS) + 1
  local p = M.PRESETS[M.preset_index]
  M.set_latency(p.delay, p.jitter)
end

function M.next_preset() M.use_preset(M.preset_index + 1) end
function M.preset_label() return M.PRESETS[M.preset_index].label end

--- Forget every socket and zero the latency — between acceptance cases.
function M.reset()
  M._live = {}
  M.delay_ms, M.jitter_ms = 0, 0
  M.preset_index = 1
end

-- A tiny LCG rather than math.random: the injector must not perturb (or be
-- perturbed by) whatever else in the process is drawing random numbers.
local function one_way()
  M._seed = (M._seed * 1103515245 + 12345) % 2147483648
  return M.delay_ms + (M._seed / 2147483648) * M.jitter_ms
end

local function enqueue(q, now, kind, payload)
  local at = now + one_way()
  if at < q.last then at = q.last end     -- the wire never reorders
  q.last = at
  q[#q + 1] = { at = at, kind = kind, payload = payload }
end

--- Wrap a room's connection. Call right after the room has joined, once its own
--- handlers are installed.
function M.wrap(room, now_ms)
  local conn = room.connection
  if not conn or conn._nd then return end

  -- reconnect built this room a NEW connection: wraps still pointing at the
  -- old one would replay their queues into the room's handlers — retire them
  for i = #M._live, 1, -1 do
    local stale = M._live[i]
    if stale.room == room and stale.conn ~= conn then table.remove(M._live, i) end
  end

  local w = {
    room = room,
    conn = conn,
    inbound = { last = 0 },
    outbound = { last = 0 },
    real_send = conn.send,
    real_emit = conn.emit,
  }
  conn._nd = w

  conn.send = function(_self, bytes)
    enqueue(w.outbound, now_ms(), "send", bytes)
  end

  local DELAYED = { message = true, close = true, error = true, open = true }
  conn.emit = function(self, event, ...)
    if not DELAYED[event] then return w.real_emit(self, event, ...) end
    enqueue(w.inbound, now_ms(), event, { n = select('#', ...), ... })
  end

  M._live[#M._live + 1] = w
  return w
end

local function drain(w, q, now)
  while #q > 0 and q[1].at <= now do
    local p = table.remove(q, 1)
    if p.kind == "send" then
      w.real_send(w.conn, p.payload)
    else
      w.real_emit(w.conn, p.kind, unpack(p.payload, 1, p.payload.n))
    end
  end
end

--- Drain everything due. Call once per frame, on the main loop.
function M.pump(now_ms)
  for i = 1, #M._live do
    local w = M._live[i]
    drain(w, w.outbound, now_ms)
    drain(w, w.inbound, now_ms)
  end
end

--- Kill every live socket uncleanly (close 4010, MAY_TRY_RECONNECT) — the
--- SDK sees a drop, not a leave, and its auto-reconnect takes over. The K key.
function M.drop()
  for i = 1, #M._live do
    local w = M._live[i]
    if w.conn.state ~= "CLOSED" then w.conn:close(4010) end
  end
end

function M.in_flight()
  local n = 0
  for i = 1, #M._live do
    n = n + #M._live[i].inbound + #M._live[i].outbound
  end
  return n
end

return M
